### ISSUE TITLE
Coldcard keystore has mainnet xpub stored for testnet wallet

### ISSUE THREAD

**Issue by jlopp** [spesmilo/electrum#6722]
Running electrum-4.0.4-x86_64.AppImage on Ubuntu 20.10

Created a 2-of-3 p2sh-p2wsh wallet with 2 coldcards (1 mk2, 1 mk3) and 1 Upub master key.

Electrum finds the funds and shows both coldcards as connected via USB. If I only have the Mk3 plugged in, I'm able to add a signature from it just fine. However, if I plug in the Mk2 it shows it as connected in the bottom right but throws an error trying to add a signature.

![Screenshot from 2020-11-09 17-23-32](https://user-images.githubusercontent.com/288011/98604138-288fca80-22b1-11eb-928d-cb910e0b924c.png)
![Screenshot from 2020-11-09 17-23-48](https://user-images.githubusercontent.com/288011/98604151-2c235180-22b1-11eb-8383-dc57893ede84.png)

<pre>
RuntimeError: Expecting cabe9eb7 but that's not what's connected?!
E | gui.qt.main_window.[dual_coldcard_recovery] | on_error
Traceback (most recent call last):
  File "/tmp/.mount_electrQ3YosC/usr/lib/python3.7/site-packages/electrum/gui/qt/util.py", line 840, in run
    result = task.task()
  File "/tmp/.mount_electrQ3YosC/usr/lib/python3.7/site-packages/electrum/wallet.py", line 1667, in sign_transaction
    k.sign_transaction(tmp_tx, password)
  File "/tmp/.mount_electrQ3YosC/usr/lib/python3.7/site-packages/electrum/plugins/coldcard/coldcard.py", line 305, in wrapper
    return func(self, *args, **kwargs)
  File "/tmp/.mount_electrQ3YosC/usr/lib/python3.7/site-packages/electrum/plugins/coldcard/coldcard.py", line 373, in sign_transaction
    client = self.get_client()
  File "/tmp/.mount_electrQ3YosC/usr/lib/python3.7/site-packages/electrum/plugins/coldcard/coldcard.py", line 286, in get_client
    rv.verify_connection(xfp_int, self.ckcc_xpub)
  File "/tmp/.mount_electrQ3YosC/usr/lib/python3.7/site-packages/electrum/plugin.py", line 362, in wrapper
    return run_in_hwd_thread(partial(func, *args, **kwargs))
  File "/tmp/.mount_electrQ3YosC/usr/lib/python3.7/site-packages/electrum/plugin.py", line 355, in run_in_hwd_thread
    return fut.result()
  File "/tmp/.mount_electrQ3YosC/usr/lib/python3.7/concurrent/futures/_base.py", line 435, in result
    return self.__get_result()
  File "/tmp/.mount_electrQ3YosC/usr/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/tmp/.mount_electrQ3YosC/usr/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/tmp/.mount_electrQ3YosC/usr/lib/python3.7/site-packages/electrum/plugins/coldcard/coldcard.py", line 105, in verify_connection
    xfp2str(expected_xfp))
RuntimeError: Expecting cabe9eb7 but that's not what's connected?!
</pre>

**Comment by jlopp** [comment]
I've also duplicated this with two mk3 coldcards running the latest firmware, so it's not a mk2 issue. With the dual mk3 setup I'm unable to successfully add a signature from either device, regardless of what order I connect them in, or even if they're both connected.

**Comment by SomberNight** [comment]
So just to be clear, if only one coldcard is connected (but the multisig contains multiple coldcard cosigners), it gets the error?

**Comment by jlopp** [comment]
Correct, and regardless of which device is plugged in I get an error that shows the fingerprint of the device that is listed first in the bottom right corner (the leftmost device.)

RuntimeError("Expecting eadeee15 but that's not what's connected?!")

Both devices are running firmware 3.1.9; the only difference I see is that the device throwing the error is running bootloader 2.0.1 while the other is running 2.0.0

**Comment by SomberNight** [comment]
Can you run and share the output of `[ks.dump() for ks in wallet.get_keystores()]` ?

**Comment by jlopp** [comment]
Well this is interesting that the device with the error shows an xpub for the ckcc_xpub while the other device shows a tpub. This is a testnet wallet.

>>> [ks.dump() for ks in wallet.get_keystores()]
[{'type': 'hardware', 'hw_type': 'coldcard', 'xpub': 'Upub5Ritb9YZEzQxmMHV78Fs7Bja3LfGkVSLeSDPVaXvUtSEo62TZPDSzfNMcX7Qk31wy7sC2a2BpEF89X7f4w3iBha6kjp7ynLZVcxRtp2dZ3t', 'derivation': 'm/49/1/6', 'root_fingerprint': 'eadeee15', 'label': 'Coldcard eadeee15', 'soft_device_id': None, 'ckcc_xpub': 'xpub661MyMwAqRbcEpQfFVUy7VxbPyAK5nRwGtPy4sjRFJb6cg5cmV3juBejpeHn2swWRdgUmBUoyF16ZUxAoLLgse93UJ3zBXUDqR4WUhp4weR'}, {'type': 'hardware', 'hw_type': 'coldcard', 'xpub': 'Upub5R5zajQhuQvQPvJUqrde7BMuT7KyJBc41c6fcxEC6LYom7iWna9N3E5DHWL53oxJabWVKXuXAvpt37BiwijtqYXgjtAMyBX5TUey8AN2Kbn', 'derivation': 'm/49/1/6', 'root_fingerprint': 'f3a9c25d', 'label': 'Coldcard f3a9c25d', 'soft_device_id': None, 'ckcc_xpub': 'tpubD6NzVbkrYhZ4Wa3rseSPcZ1TEunhfdqynmZLrwET6JDhZy6cC4myfMvhLjXDxLgR73Su7hwK1nWE78bZ2Sk1Agd5A1u9RgBbURemS6cCTuE'}, {'type': 'bip32', 'pw_hash_version': 1, 'xpub': 'Upub5QaiwZWYcoHCGjGdy1xgWL6A3J9fWLMFMSYgppCq9ERhwL6vMjAUc3VNQdKc9PFimvqA1AMsKVY5pG9D4THmuM5zapHkjEWXiYGsyRAUnE6', 'xprv': None, 'derivation': None, 'root_fingerprint': None}]

**Comment by SomberNight** [comment]
so on this line
https://github.com/spesmilo/electrum/blob/fa8c751abfbb83f16b4ef69d77b8ac05b424b2fb/electrum/plugins/coldcard/coldcard.py#L102-L103
does it try to compare testnet against mainnet?

**Comment by jlopp** [comment]
```
I/i | interface.[hsmithsxurybd7uh.onion:53012] | disconnecting due to: ErrorGettingSSLCertFromServer(ConnectError(gaierror(-2, 'Name or service not known')))
I/n | network | couldn't launch iface hsmithsxurybd7uh.onion:53012:s -- CancelledError()
I/i | interface.[wgzwu3vg3i54wayx.onion:51002] | disconnecting due to: ErrorGettingSSLCertFromServer(ConnectError(gaierror(-2, 'Name or service not known')))
I/n | network | couldn't launch iface wgzwu3vg3i54wayx.onion:51002:s -- CancelledError()
I/i | interface.[tn.not.fyi:55002] | disconnecting due to GracefulDisconnect('too many connected servers already in bucket 50.2.0.0/16')
I/n | network | couldn't launch iface tn.not.fyi:55002:s -- CancelledError()
I | plugin.DeviceMgr | getting client for keystore
I | plugin.DeviceMgr | scanning devices...
D | util.profiler | DeviceMgr.scan_devices 0.0554
I | plugin.DeviceMgr | end client for keystore
I | plugin.DeviceMgr | getting client for keystore
I | plugin.DeviceMgr | scanning devices...
D | util.profiler | DeviceMgr.scan_devices 0.0267
I | plugin.DeviceMgr | end client for keystore
I | plugin.DeviceMgr | getting client for keystore
I | plugin.DeviceMgr | scanning devices...
D | util.profiler | DeviceMgr.scan_devices 0.0502
I | plugin.DeviceMgr | end client for keystore
I | plugins.coldcard.coldcard | xpubs. reported by device: tpubD6NzVbkrYhZ4WcbWhgU4KRHxj6Fy5AvzpWzGRRnZbDLzMxkKqhtq3dKa4gNZZNtkDYDPNkzgWLq9FfSnCCBvm5bLXtoHs48TRNevEmFtvCz. stored in file: xpub661MyMwAqRbcEpQfFVUy7VxbPyAK5nRwGtPy4sjRFJb6cg5cmV3juBejpeHn2swWRdgUmBUoyF16ZUxAoLLgse93UJ3zBXUDqR4WUhp4weR
E | gui.qt.main_window.[dual_coldcard_mk3_recovery] | on_error
Traceback (most recent call last):
  File "/tmp/.mount_electrPWkzRs/usr/lib/python3.7/site-packages/electrum/gui/qt/util.py", line 840, in run
    result = task.task()
  File "/tmp/.mount_electrPWkzRs/usr/lib/python3.7/site-packages/electrum/wallet.py", line 1667, in sign_transaction
    k.sign_transaction(tmp_tx, password)
  File "/tmp/.mount_electrPWkzRs/usr/lib/python3.7/site-packages/electrum/plugins/coldcard/coldcard.py", line 305, in wrapper
    return func(self, *args, **kwargs)
  File "/tmp/.mount_electrPWkzRs/usr/lib/python3.7/site-packages/electrum/plugins/coldcard/coldcard.py", line 373, in sign_transaction
    client = self.get_client()
  File "/tmp/.mount_electrPWkzRs/usr/lib/python3.7/site-packages/electrum/plugins/coldcard/coldcard.py", line 286, in get_client
    rv.verify_connection(xfp_int, self.ckcc_xpub)
  File "/tmp/.mount_electrPWkzRs/usr/lib/python3.7/site-packages/electrum/plugin.py", line 362, in wrapper
    return run_in_hwd_thread(partial(func, *args, **kwargs))
  File "/tmp/.mount_electrPWkzRs/usr/lib/python3.7/site-packages/electrum/plugin.py", line 355, in run_in_hwd_thread
    return fut.result()
  File "/tmp/.mount_electrPWkzRs/usr/lib/python3.7/concurrent/futures/_base.py", line 435, in result
    return self.__get_result()
  File "/tmp/.mount_electrPWkzRs/usr/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/tmp/.mount_electrPWkzRs/usr/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/tmp/.mount_electrPWkzRs/usr/lib/python3.7/site-packages/electrum/plugins/coldcard/coldcard.py", line 105, in verify_connection
    xfp2str(expected_xfp))
RuntimeError: Expecting eadeee15 but that's not what's connected?!

```

**Comment by jlopp** [comment]
It's /possible/ that I set up this multisig wallet with one of the coldcard "blockchain" setting on mainnet and then later switched it to testnet, but I honestly don't remember the specific sequence of events.

I just set up the same wallet again after checking that both coldcards are set to testnet and I'm able to sign transactions.

**Comment by SomberNight** [comment]
> It's /possible/ that I set up this multisig wallet with one of the coldcard "blockchain" setting on mainnet and then later switched it to testnet, but I honestly don't remember the specific sequence of events.

Yes indeed I can reproduce that way.

**Comment by Rushmeister909** [comment]
@SomberNight  
Hi SomberNight,

I have this exact same error. I imported a mainnet xpub in my coldcard mk3. I use the latest firmware of coldcard : 2021-01-14T1617-v3.2.2 and electrum 4.0.9

If i try to sign a transaction it throws this error in Electrum. 
If I sign a transaction via another wallet were I connect my hww to, it works just fine. 

Any ideas what to do or how to prevent this?

**Comment by SomberNight** [comment]
@Rushmeister909 Just recreate the wallet file with the correct xpubs this time.

Did you already use Electrum 4.0.9 when the wallet file was created, or did you upgrade after that?
It should no longer be possible to reproduce this issue with recent versions (but existing already corrupted wallet files might be out there).

**Comment by Rushmeister909** [comment]
Hi @SomberNight  Thanks alot for the quick reply.

I created this wallet file for the first time with my Mk3. Before that it was not existing. 
however, the keys are imported and not created on the Mk3. I dont know if that could cause any issue?
 
I did only use electrum 4.0.9 before with a Ledger and not with my MK3. 

I will remove the wallet files that were created with my MK3 and will try it again. I will report back if that succeeds. 
Thanks!
